packages/core/src/storage/namespace.ts

```diff
@@ -0,0 +1,305 @@
+      throw new DuplicateNamespaceError(id, contractName, srcs);
+    } else {
+      // The src locations of the namespace structs are no longer needed at this point, so remove them
+      namespaces[id] = namespace;
+    }
+  }
+  layout.namespaces = namespaces;
+}
+
+/**
+ * Load namespaces (including loading recursive type information) and original source locations.
+ */
+function loadNamespacesWithSrcs(
+  decodeSrc: SrcDecoder,
+  layout: StorageLayout,
+  origContext: CompilationContext,
+  namespacedContext?: CompilationContext,
+) {
+  const result: Record<string, NamespaceWithSrcs> = {};
+
+  const origLinearizedIds = origContext.contractDef.linearizedBaseContracts;
+  if (namespacedContext === undefined) {
+    for (let i = 0; i < origLinearizedIds.length; i++) {
+      const origReferencedContract = getReferencedContract(origContext, origLinearizedIds[i]);
+
+      const context = { ...origContext, contractDef: origReferencedContract };
+      addContractNamespacesWithSrcs(result, decodeSrc, layout, context, origReferencedContract);
+    }
+  } else {
+    const namespacedLinearizedIds = namespacedContext.contractDef.linearizedBaseContracts;
+    assert(origLinearizedIds.length === namespacedLinearizedIds.length);
+    for (let i = 0; i < origLinearizedIds.length; i++) {
+      const origReferencedContract = getReferencedContract(origContext, origLinearizedIds[i]);
+      const namespacedReferencedContract = getReferencedContract(namespacedContext, namespacedLinearizedIds[i]);
+
+      const context = { ...namespacedContext, contractDef: namespacedReferencedContract };
+      addContractNamespacesWithSrcs(result, decodeSrc, layout, context, origReferencedContract);
+    }
+  }
+
+  return result;
+}
+
+/**
+ * Gets the contract definition for the given referenced id.
+ */
+function getReferencedContract(context: CompilationContext, referencedId: number) {
+  // Optimization to avoid dereferencing if the referenced id is the same as the current contract
+  return context.contractDef.id === referencedId
+    ? context.contractDef
+    : context.deref(['ContractDefinition'], referencedId);
+}
+
+/**
+ * Add namespaces and source locations for the given compilation context's contract.
+ * Does not include inherited contracts.
+ */
+function addContractNamespacesWithSrcs(
+  namespaces: Record<string, NamespaceWithSrcs>,
+  decodeSrc: SrcDecoder,
+  layout: StorageLayout,
+  context: CompilationContext,
+  origContractDef: ContractDefinition,
+) {
+  for (const node of context.contractDef.nodes) {
+    if (isNodeType('StructDefinition', node)) {
+      const storageLocationArg = getStorageLocationArg(node);
+      if (storageLocationArg !== undefined) {
+        const origSrc = decodeSrc(getOriginalStruct(node.canonicalName, origContractDef));
+
+        if (namespaces[storageLocationArg] !== undefined) {
+          namespaces[storageLocationArg].srcs.push(origSrc);
```

Comment:
Is there a reason why we don't just throw an error at this point, and instead do that later at `checkAndSaveLayout`?

Reply:
Doing it later allows us to collect all of the conflicts and report them all at once (in case there are more than 2 of the same namespace)